mono/utils/mono-state.c

```diff
@@ -31,266 +31,280 @@ extern GCStats mono_gc_stats;
 #include <mach/task_info.h>
 #endif
 
+#define FORMAT_APPEND_STATIC(writer, ...) do { \
```

Comment:
Why a macro? This looks largely used like `mono_json_writer_printf`, so how about `mono_state_writer_printf` as a function.

Reply:
Agreed.

One thing gained by a macro is the assert is reported in-situ. But agreed -- stack trace doubles for assert context, most people don't think about assert context anyway, and  the primary reason to favor macros over function is "syntactic structure within macro", like the error handling and error_init macros. Also almost-guaranteed inlining, which doesn't apply here.  Or to double as C++ templates, which doesn't apply here.

glib has something like this -- g_string_append_printf -- that should  be reusable. It might need some "policy" flag to inhibit realloc.